breadbox/breadbox/service/metadata.py

```diff
@@ -0,0 +1,71 @@
+import logging
+from typing import Any, Optional
+
+from breadbox.crud import dataset as dataset_crud
+from breadbox.crud import types as types_crud
+from breadbox.db.session import SessionWithUser
+from breadbox.schemas.custom_http_exception import ResourceNotFoundError
+from breadbox.models.dataset import MatrixDataset, TabularDataset
+
+from depmap_compute.slice import SliceQuery
+
+log = logging.getLogger(__name__)
+
+
+def get_tabular_dataset_metadata_annotations(
+    db: SessionWithUser, dataset: MatrixDataset, metadata_col_name: str,
+) -> dict[str, Any]:
+    """
+    For the given tabular dataset, load a column from the associated metadata.
+    Return a dictionary of values keyed by given id.
+    Note: this should work regardless of whether or not the given dataset IS metadata
+    or HAS metadata.
+    """
+    full_metadata_col = types_crud.get_dimension_type_metadata_col(
+        db, dimension_type_name=dataset.sample_type, col_name=metadata_col_name
+    )
+    # Filter the metadata to only include the given IDs belonging to this dataset
+    dataset_samples = dataset_crud.get_dataset_columns(db, dataset)
+    dataset_given_ids = [sample.given_id for sample in dataset_samples]
+    return {given_id: full_metadata_col[given_id] for given_id in dataset_given_ids}
+
+
+def get_tabular_dataset_labels_by_id(
+    db: SessionWithUser, dataset: TabularDataset,
+) -> dict[str, str]:
+    """
+    For the index (rows) of a tabular dataset, try loading labels from metadata.
+    If there are no labels in the metadata or there is no metadata, then just return the given IDs as labels.
+    """
+    metadata_labels = get_tabular_dataset_metadata_annotations(
+        db=db, dataset=dataset, metadata_col_name="label"
+    )
+    if metadata_labels:
+        return metadata_labels
+    else:
+        columns = dataset_crud.get_dataset_columns(db=db, dataset=dataset)
+        return {column.given_id: column.given_id for column in columns}
```

Comment:
Oh actually, after I opened this PR, I realized that this particular case is going to cause issues because it's loading the given IDs by column, not by row. I'll work on fixing this and then let you know when it's actually ready for review. Sorry about that!

Reply:
Okay sorry for the confusion, this should actually be ready to review now. 

@jessica-cheng I added you as a reviewer because I think you're the most familiar with the existing tabular dataset stuff (I'm relatively new to this part of breadbox and was learning my way around the existing crud functions we have). Any feedback is welcome!